### PR TITLE
Bug/107057 ipo export til excel performance tuning second attempt

### PR DESCRIPTION
**Pull request by trondlindanger**
[AB#107057](https://statoildeveloper.visualstudio.com/b6d97460-2e53-47ae-aad4-b41396d90828/_workitems/edit/107057)

### CODE REVIEW

**Comment by janid1967** [comment]
Build + running tests fail @trondlindanger

File: src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Invitation.cs [hunk 1]
```diff
@@ -147,6 +151,11 @@ public void AddParticipant(Participant participant)
                 throw new ArgumentException($"Can't relate item in {participant.Plant} to item in {Plant}");
             }
 
+            if (_participants.Contains(participant))
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Invitation.cs:154]: What is the motivation here? Is it to prohibit that the very same participant-object is not added twice?
Or is it to prohibit that two different participant-object, which represent the same end user is not added twice?

File: src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/CommPkg.cs [hunk 2]
```diff
@@ -55,6 +56,9 @@ protected CommPkg()
             Guid = guid;
         }
 
+        [NotMapped] // EF defines a shadow property for InvitationId.
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/CommPkg.cs:59]: I don't like this .... this will cause misunderstanding since the very same class will be used in normal EF context, and then this property will never have value from DB. I think we must discuss other possible solutions with team. I stop review here for now

Comment (janid1967) [reply src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/CommPkg.cs:59]: I have an idea on other solution

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:1]: Remove unused usings

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/IExportIpoRepository.cs
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/IExportIpoRepository.cs:1]: Remove unused usings

File: src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Invitation.cs [hunk 3]
```diff
@@ -135,6 +135,10 @@ public void RemoveAttachment(Attachment attachment)
             AddDomainEvent(new AttachmentRemovedEvent(Plant, Guid, attachment.FileName));
         }
 
+        public void AddParticipants(IReadOnlyCollection<Participant> participants) => _participants.AddRange(participants);
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Invitation.cs:138]: Can't do! Should loop input participants, and call AddParticipant to execute business-logic there!
Same for AddCommPkgs and AddMcPkgs below

So, if you move these 3 methods to another derived class for Inivitation, AddParticipant should still be used (which again prohibir expose _participants as protected)

File: src/Equinor.ProCoSys.IPO.WebApi/Excel/ExcelConverter.cs [hunk 4]
```diff
@@ -4,13 +4,15 @@
...
+using Microsoft.Extensions.Logging;
+using Microsoft.Identity.Client;
 using Color = System.Drawing.Color;
 
 namespace Equinor.ProCoSys.IPO.WebApi.Excel
 {
     public class ExcelConverter : IExcelConverter
     {
-        public MemoryStream Convert(ExportDto dto)
+        public MemoryStream Convert(ExportDto dto, ILogger logger)
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.WebApi/Excel/ExcelConverter.cs:15]: why not create the logger by DI as all other places?

File: src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/InvitationsController.cs
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/InvitationsController.cs:1]: Undo changes in this file and create logger in src/Equinor.ProCoSys.IPO.WebApi/Excel/ExcelConverter.cs by DI (as here)

File: src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Invitation.cs (hunk 1 above)
Comment (trondlindanger) [reply src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Invitation.cs:154]: To avoid that the same object is added twice.

File: src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Invitation.cs (hunk 3 above)
Comment (trondlindanger) [reply src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Invitation.cs:138]: Looping through 650000 participants... will check again which performance impact that has.

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/DapperRepositoryBase.cs [hunk 5]
```diff
@@ -10,13 +10,13 @@ namespace Equinor.ProCoSys.IPO.Infrastructure.Repositories;
 
 public class DapperRepositoryBase
 {
-    private readonly IPOContext _context;
+    public readonly IPOContext Context;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/DapperRepositoryBase.cs:13]: Add a comment why this should be private, but we need to be make it public. Start the comment with phrase // HACK

File: src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/CommPkg.cs
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/CommPkg.cs:1]: Undo all changes in this file, line 9, 59, 60 and 61
The file should not appear as changed here in PR

File: src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/McPkg.cs [hunk 6]
```diff
@@ -6,6 +6,7 @@
 using Equinor.ProCoSys.Common.Time;
 using Equinor.ProCoSys.Common;
 using Equinor.ProCoSys.IPO.Domain.AggregateModels.CertificateAggregate;
+using System.ComponentModel.DataAnnotations.Schema;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/McPkg.cs:9]: Undo all changes in this file, line 9, 65 and 66
The file should not appear as changed here in PR

File: src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Participant.cs [hunk 7]
```diff
@@ -1,4 +1,5 @@
 ﻿using System;
+using System.ComponentModel.DataAnnotations.Schema;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Participant.cs:2]: Undo changes in this file, line 2, 46 and 47. OK to keep the reformatting of constrctor

File: src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/DapperCommPkg.cs
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/DapperCommPkg.cs:1]: This comment goes for all classes made here named DapperXXXX
1) Use file-scoped namespace (this is a setting in Visual Studio, where you decide that file-scoped namespace are made in new files)
1) Remove the using and remove [NotMapped] with it's comment
1) Move the file from Equinor.ProCoSys.IPO.Domain.AggregateModels.InvitationAggregate to Equinor.ProCoSys.IPO.Infrastructure.Repositories.ExportIPOs. Remember to update namespace

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
+﻿using System.Collections.Generic;
+using System.Data;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Transactions;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:6]: remove unused using

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+using Dapper;
+using Equinor.ProCoSys.Common.Misc;
+using Equinor.ProCoSys.IPO.Domain.AggregateModels.InvitationAggregate;
+using Microsoft.EntityFrameworkCore;
+
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:11]: remove extra blank line here and line 14

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    public async Task<List<Invitation>> GetInvitationsWithIncludesAsync(
+        List<int> invitationIds,
+        IPlantProvider plantProvider,
+        CancellationToken cancellationToken)
+    {
+        await using (var connection = Context.Database.GetDbConnection())
+        {
+            await connection.OpenAsync(cancellationToken);
+            await using (var transaction = await connection.BeginTransactionAsync())
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:28]: BeginTransactionAsync(cancellationToken))

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+            {
+                transaction.Commit();
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:30]: await transaction.CommitAsync(cancellationToken);

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+                var parameters = new { plant = plantProvider.Plant };
+                using (var results = await connection.QueryMultipleAsync("GetInvitations", parameters, commandType: CommandType.StoredProcedure))
+                {
+                    var invitations = await results.ReadAsync<Invitation>();
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:34]: Either `var invitations = (await results.ReadAsync<Invitation>()).ToList();`
or create a new var invitationList = invitations.ToList();

simplify the return invitations.ToList() on line 72 to either return invitations; or return invitationList; accordantly

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+                        var participantsForThisInvitation = participantsByInvitation[participantByInvitation.Key].AsList();
+                        participantsForThisInvitation.ForEach(p =>
+                        {
+                            invitationsLookup[participantByInvitation.Key].First().AddParticipant(p);
+                        });
+                    }
+
+                    foreach (var commPkgByInvitation in commPkgsByInvitation)
+                    {
+                        var commpkgs = commPkgsByInvitation[commPkgByInvitation.Key].AsList();
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:57]: rename to commPkgs

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+                        commpkgs.ForEach(c => {
+                            invitationsLookup[commPkgByInvitation.Key].First().AddCommPkg(c);
+                        });
+                    }
+
+                    foreach (var mcPkgByInvitation in mcPkgsByInvitation)
+                    {
+                        var mcpkgs = mcPkgsByInvitation[mcPkgByInvitation.Key].AsList();
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:65]: rename to mcPkgs

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+                            invitationsLookup[mcPkgByInvitation.Key].First().AddMcPkg(m);
+                        });
+                    }
+
+                    return invitations.ToList();
+                }
+            }
+        }
+
+        return null;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:77]: remove the return null; ...code flow never came here

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/IExportIpoRepository.cs [hunk 16]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/IExportIpoRepository.cs:1]: remove unused using + extra blank line at end of file

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+        CancellationToken cancellationToken)
+    {
+        await using (var connection = Context.Database.GetDbConnection())
+        {
+            await connection.OpenAsync(cancellationToken);
+            await using (var transaction = await connection.BeginTransactionAsync())
+            {
+                transaction.Commit();
+                var parameters = new { plant = plantProvider.Plant };
+                using (var results = await connection.QueryMultipleAsync("GetInvitations", parameters, commandType: CommandType.StoredProcedure))
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:32]: await using

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+public class ExportIpoRepository : DapperRepositoryBase, IExportIpoRepository
+{
+    public ExportIpoRepository(IPOContext context) : base(context) { }
+
+    public async Task<List<Invitation>> GetInvitationsWithIncludesAsync(
+        List<int> invitationIds,
+        IPlantProvider plantProvider,
+        CancellationToken cancellationToken)
+    {
+        await using (var connection = Context.Database.GetDbConnection())
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:25]: for all using statements in this class, convert to using-declarations. As this:
image

File: src/Equinor.ProCoSys.IPO.Infrastructure/Migrations/IPOContextModelSnapshot.cs [hunk 19]
```diff
@@ -239,8 +239,7 @@ protected override void BuildModel(ModelBuilder modelBuilder)
                     b.Property<Guid>("Guid")
                         .HasColumnType("uniqueidentifier");
 
-                    b.Property<int?>("InvitationId")
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Migrations/IPOContextModelSnapshot.cs:242]: ohh .. I guess this change has been added since you during the process had the InvitationId as MAPPED property 
We should compare our schema with my schema (since I've not run the migration yet)

I'm pretty sure the solution is to revert this change, but we must be sure .. need to talk about it

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+using Dapper;
+using Equinor.ProCoSys.Common.Misc;
+using Equinor.ProCoSys.IPO.Domain.AggregateModels.InvitationAggregate;
+using Microsoft.EntityFrameworkCore;
+
+
+namespace Equinor.ProCoSys.IPO.Infrastructure.Repositories.ExportIPOs;
+
+
+public class ExportIpoRepository : DapperRepositoryBase, IExportIpoRepository
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:16]: consider adding unit tests, if possible. I see that OutstandingIpoRepository has some tests, building a context using sqllite. Maybe create a PBI for next sprint and leave a todo here with // Todo <PBI-number> Create unit tests

File: src/Equinor.ProCoSys.IPO.Query/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExcelQueryHandler.cs [hunk 21]
```diff
@@ -66,13 +74,21 @@ public async Task<Result<ExportDto>> Handle(GetInvitationsForExportQuery request
         private async Task<List<Invitation>> GetOrderedInvitationsWithIncludesAsync(GetInvitationsForExportQuery request,
             CancellationToken cancellationToken)
         {
+            _logger.LogInformation($"Export to excel. Creating queryable with filter...");
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Query/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExcelQueryHandler.cs:77]: remove superfluous $ on 5 _logger.LogInformation($" in this method

File: src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/InvitationsController.cs [hunk 22]
```diff
@@ -48,13 +49,16 @@ namespace Equinor.ProCoSys.IPO.WebApi.Controllers.Invitation
     [Route("Invitations")]
     public class InvitationsController : ControllerBase
     {
+        private readonly ILogger<InvitationsController> _logger;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.WebApi/Controllers/Invitation/InvitationsController.cs:52]: why keep the logger here now!! Undo all changes in in this file. The file should not appear as changed in PR

File: src/Equinor.ProCoSys.IPO.WebApi/Excel/ExcelConverter.cs [hunk 23]
```diff
@@ -3,13 +3,22 @@
 using System.IO;
 using System.Linq;
 using Equinor.ProCoSys.IPO.Query.GetInvitationsQueries.GetInvitationsForExport;
+using Equinor.ProCoSys.IPO.WebApi.Controllers.Invitation;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.WebApi/Excel/ExcelConverter.cs:6]: remove unused usings

File: src/Equinor.ProCoSys.IPO.WebApi/Excel/ExcelConverter.cs [hunk 24]
```diff
@@ -3,13 +3,22 @@
...
+using Microsoft.Identity.Client;
 using Color = System.Drawing.Color;
 
 namespace Equinor.ProCoSys.IPO.WebApi.Excel
 {
     public class ExcelConverter : IExcelConverter
     {
+        private readonly ILogger<ExcelConverter> _logger;
+
+        public ExcelConverter(
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.WebApi/Excel/ExcelConverter.cs:19]: make line 19 + 20 to one line

File: src/Equinor.ProCoSys.IPO.WebApi/Excel/ExcelConverter.cs [hunk 25]
```diff
@@ -22,18 +31,28 @@ public MemoryStream Convert(ExportDto dto)
 
             try
             {
+                _logger.LogInformation($"Export to excel. Start generating sheets...");
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.WebApi/Excel/ExcelConverter.cs:34]: remove superfluous $

File: src/Equinor.ProCoSys.IPO.WebApi/Excel/ExcelConverter.cs [hunk 26]
```diff
@@ -22,18 +31,28 @@ public MemoryStream Convert(ExportDto dto)
...
                     GenerateParticipantsSheet(xlsxWriter, invitationsHeader, exportInvitationDtos, dateStyle,
                         normalStyle);
+                    _logger.LogInformation("Export to excel. Generating history sheet...");
+
                     GenerateHistorySheet(xlsxWriter, normalStyle, invitationsHeader, exportInvitationDtos, dateStyle);
                 }
+                _logger.LogInformation("Export to excel. Completed.");
 
                 return stream;
             }
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.WebApi/Excel/ExcelConverter.cs:58]: remove the superfluous catch below

File: src/Equinor.ProCoSys.IPO.WebApi/Excel/IExcelConverter.cs [hunk 27]
```diff
@@ -1,5 +1,6 @@
 ﻿using System.IO;
 using Equinor.ProCoSys.IPO.Query.GetInvitationsQueries.GetInvitationsForExport;
+using Microsoft.Extensions.Logging;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.WebApi/Excel/IExcelConverter.cs:3]: Revert changes!

File: src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs [hunk 28]
```diff
@@ -1,24 +1,33 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Threading;
 using System.Threading.Tasks;
 using Equinor.ProCoSys.Common.Misc;
+using Equinor.ProCoSys.IPO.Command.InvitationCommands.CancelPunchOut;
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs:7]: remove

File: src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs [hunk 29]
```diff
@@ -1,24 +1,33 @@
...
     [TestClass]
     public class GetInvitationsForExportQueryHandlerTests : ReadOnlyTestsBaseInMemory
     {
+        private ILogger<GetInvitationsForExportQuery> _logger;
+        private IExportIpoRepository _exportIpoRepository;
+        private Mock<IExportIpoRepository> _exportIpoRepositoryMock;
+
         private Invitation _invitation1;
         private Invitation _invitation2;
         private Invitation _invitation3;
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs:33]: remove unused int _participantId1; at line 34

File: src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs [hunk 30]
```diff
@@ -376,35 +399,6 @@ public async Task HandleGetInvitationsForExportQuery_ShouldFilterOnStatusPlanned
             }
         }
 
-        [TestMethod]
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs:379]: Why remove this particular test?

File: src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs [hunk 31]
```diff
@@ -1,24 +1,33 @@
...
 using ServiceResult;
 
 namespace Equinor.ProCoSys.IPO.Query.Tests.GetInvitationsQueries.GetInvitationsForExport
 {
     [TestClass]
     public class GetInvitationsForExportQueryHandlerTests : ReadOnlyTestsBaseInMemory
     {
+        private ILogger<GetInvitationsForExportQuery> _logger;
+        private IExportIpoRepository _exportIpoRepository;
+        private Mock<IExportIpoRepository> _exportIpoRepositoryMock;
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs:29]: remove _exportIpoRepositoryMock, see comment in Setup

File: src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs [hunk 32]
```diff
@@ -47,6 +56,20 @@ public class GetInvitationsForExportQueryHandlerTests : ReadOnlyTestsBaseInMemor
         private readonly Project _project1 = new(TestPlant, _projectName, $"Description of {_projectName}", _project1Guid);
         private readonly Project _project2 = new(TestPlant, _projectName2, $"Description of {_projectName2}" , _project2Guid);
 
+        [TestInitialize]
+        public void Setup()
+        {
+            _logger = new Mock<ILogger<GetInvitationsForExportQuery>>().Object;
+            _exportIpoRepositoryMock = new Mock<IExportIpoRepository>();
```
Comment (janid1967) [inline src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs:63]: rename and make local var exportIpoRepositoryMock

**Review by janid1967 — CHANGES_REQUESTED** [review]
See comments

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs (hunk 20 above)
Comment (trondlindanger) [reply src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:16]: Here is a PBI for this: https://dev.azure.com/statoildeveloper/Statoil.NextGenProCoSys/_workitems/edit/109937

File: src/Equinor.ProCoSys.IPO.Infrastructure/Migrations/IPOContextModelSnapshot.cs (hunk 19 above)
Comment (trondlindanger) [reply src/Equinor.ProCoSys.IPO.Infrastructure/Migrations/IPOContextModelSnapshot.cs:242]: Right. So what do we need to compare here? An export of the loca data model?

File: src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs (hunk 30 above)
Comment (trondlindanger) [reply src/tests/Equinor.ProCoSys.IPO.Query.Tests/GetInvitationsQueries/GetInvitationsForExport/GetInvitationsForExportQueryHandlerTests.cs:379]: As far as I understand, the test does a change to an InMemory database which is assumed to be reflected in the results. This is not the case for stored procedure since this is not supported in InMemory database as far as I can see.

File: src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/CommPkg.cs [hunk 33]
```diff
@@ -6,6 +6,7 @@
 using Equinor.ProCoSys.Common.Time;
 using Equinor.ProCoSys.Common;
 using Equinor.ProCoSys.IPO.Domain.AggregateModels.CertificateAggregate;
+using System.ComponentModel.DataAnnotations.Schema;
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/CommPkg.cs:9]: Remove. The file should not appear as changed in PR

File: src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/McPkg.cs [hunk 34]
```diff
@@ -61,7 +61,6 @@ protected McPkg()
             CommPkgGuid = commPkgGuid;
         }
 
-        // private setters needed for Entity Framework
```
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/McPkg.cs:64]: Re-add the very same comment. The file should not appear as changed in PR

File: src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Participant.cs (hunk 7 above)
Comment (janid1967) [inline src/Equinor.ProCoSys.IPO.Domain/AggregateModels/InvitationAggregate/Participant.cs:2]: Remove

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs (hunk 20 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:13]: Add a comment above the method: // Todo 109937 Add unit tests

File: src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs (hunk 12 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.IPO.Infrastructure/Repositories/ExportIPOs/ExportIpoRepository.cs:32]: No action taken since last PR. Still possible multiple enumeration waring on invitations variable:
image

**Review by janid1967 — CHANGES_REQUESTED** [review]
See comments
Some resolved comments are re-opened

(I'll try to fix the removed test in GetInvitationsForExportQueryHandlerTests. We can discuss to commit it or not, if I manage to make it before you've fixed comments)

**Comment by janid1967** [comment]
I'll fix my own comments and merge
